### PR TITLE
WIP: Support deploying and publishing plugins

### PR DESCRIPTION
**Pull request by oeway**

### CODE REVIEW

File: imjoy/imjoyPluginEngine.py [hunk 1]
```diff
@@ -477,6 +487,10 @@ def killPlugin(pid):
         if "sid" in plugins[pid]:
             if plugins[pid]["sid"] in plugin_sids:
                 del plugin_sids[plugins[pid]["sid"]]
+        if "publish_id" in plugins[pid] and plugins[pid]["publish_id"] is not None:
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]: This block seems to repeat the new block above. Can we extract it to a helper function?

File: imjoy/imjoyPluginEngine.py [hunk 2]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
 @sio.on("init_plugin", namespace=NAME_SPACE)
 async def on_init_plugin(sid, kwargs):
     try:
-        if sid in registered_sessions:
+        logger.info("init plugin %s", kwargs)
+        pid = kwargs["id"]
+        role = kwargs.get("role", "admin")
+        publish_id = kwargs.get("publish_id", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
        publish_id = kwargs.get("publish_id")
```
`None` is the default value returned if the key is missing in the dict.

File: imjoy/imjoyPluginEngine.py [hunk 3]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
...
+        pname = kwargs.get("name", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
        pname = kwargs.get("name")
```

File: imjoy/imjoyPluginEngine.py [hunk 4]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
...
-            "initialize the plugin. name=%s, id=%s, cmd=%s, workspace=%s",
-            pname,
-            id,
-            cmd,
-            workspace,
-        )
-
-        plugin_signature = "{}/{}/{}".format(workspace, pname, tag)
+            config = kwargs.get("config", {})
+            env = config.get("env", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
            env = config.get("env")
```

File: imjoy/imjoyPluginEngine.py [hunk 5]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
...
+                    #     {"type": "initialized", "dedicatedThread": True})
+                    return {
+                        "success": True,
+                        "resumed": True,
+                        "initialized": True,
+                        "secret": secret,
+                        "work_dir": os.path.abspath(work_dir),
+                    }
+        elif role == "user" and publish_id is not None:
+            ptoken = kwargs.get("plugin_token", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
            ptoken = kwargs.get("plugin_token")
```

File: imjoy/imjoyPluginEngine.py [hunk 6]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
...
+            if publish_id in publications:
+                if not publications[publish_id]["require_token"] or (
+                    publications[publish_id]["token"]
+                    and ptoken == publications[publish_id]["token"]
+                ):
+                    plugin_list = publications[publish_id]["plugin_list"]
+                    config = kwargs.get("config", {})
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]: We don't seem to use the config?

File: imjoy/imjoyPluginEngine.py [hunk 7]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
...
+                        "work_dir": os.path.abspath(work_dir),
+                    }
+        elif role == "user" and publish_id is not None:
+            ptoken = kwargs.get("plugin_token", None)
+            if publish_id in publications:
+                if not publications[publish_id]["require_token"] or (
+                    publications[publish_id]["token"]
+                    and ptoken == publications[publish_id]["token"]
+                ):
+                    plugin_list = publications[publish_id]["plugin_list"]
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]: We don't seem to use the plugin_list?

File: imjoy/imjoyPluginEngine.py [hunk 8]
```diff
@@ -1009,6 +1060,112 @@ def scandir(path, type=None, recursive=False):
...
 
 
+@sio.on("publish_plugin", namespace=NAME_SPACE)
+async def on_publish_plugin(sid, kwargs):
+    logger.info("publish plugin: %s", kwargs)
+    if sid not in registered_sessions:
+        logger.debug("client %s is not registered.", sid)
+        return {"success": False, "error": "client has not been registered."}
+
+    pid = kwargs.get("id", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
    pid = kwargs.get("id")
```

File: imjoy/imjoyPluginEngine.py [hunk 9]
```diff
@@ -1009,6 +1060,112 @@ def scandir(path, type=None, recursive=False):
...
+    name = kwargs.get("name", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
    name = kwargs.get("name")
```

File: imjoy/imjoyPluginEngine.py [hunk 10]
```diff
@@ -1009,6 +1060,112 @@ def scandir(path, type=None, recursive=False):
...
+    plugin_list = kwargs.get("plugin_list", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
    plugin_list = kwargs.get("plugin_list")
```

File: imjoy/imjoyPluginEngine.py [hunk 11]
```diff
@@ -1009,6 +1060,112 @@ def scandir(path, type=None, recursive=False):
...
+    publish_id = kwargs.get("publish_id", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
    publish_id = kwargs.get("publish_id")
```

File: imjoy/imjoyPluginEngine.py [hunk 12]
```diff
@@ -1009,6 +1060,112 @@ def scandir(path, type=None, recursive=False):
...
+    require_token = kwargs.get("require_token", None)
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]:
```suggestion
    require_token = kwargs.get("require_token")
```

File: imjoy/imjoyPluginEngine.py [hunk 13]
```diff
@@ -683,50 +697,84 @@ def connect(sid, environ):
 @sio.on("init_plugin", namespace=NAME_SPACE)
 async def on_init_plugin(sid, kwargs):
```
Comment (MartinHjelmare) [inline imjoy/imjoyPluginEngine.py]: This function is long... :see_no_evil:

File: imjoy/imjoyPluginEngine.py (hunk 1 above)
Comment (oeway) [reply imjoy/imjoyPluginEngine.py]: I agree, this file need to be refactored, there are lots of places like that.
I would like to construct classes, and use database to store all these.

File: imjoy/imjoyPluginEngine.py (hunk 2 above)
Comment (oeway) [reply imjoy/imjoyPluginEngine.py]: I see.

File: imjoy/imjoyPluginEngine.py (hunk 13 above)
Comment (oeway) [reply imjoy/imjoyPluginEngine.py]: Yes, it's very long, I hesitate to do it right now, will have to rewrite later.

**Comment by oeway** [comment]
This PR is outdated.

One thing blocking this to proceed is that when multiple client is connected to the same plugin worker, we don't know which client initiated the plugin api call. We need to think a bit more on how to do this.

**Pull request closed without merging**